File: InfoVis/scripts/terraclimate/gistemp.py
import numpy as np
import xarray as xr
import json

file = f'../../data/gistemp/gistemp1200_GHCNv4_ERSSTv5.nc'
ds = xr.open_dataset(file)

parsed = {}

for year in range(1950, 1980):
  print(f'Processing {year}')
  parsed = []

  for month in range(1, 13):
    mdata = []
    for lat in range(-89, 89, 2):
      ldata = []
      for lon in range(-179, 179, 2):
        da = ds.sel(
          time=f'{year}-{month:02}',
          lat=lat,
          lon=lon
        ).tempanomaly.values

        if len(da) != 1:
          raise Exception("data must only have one value")

        if np.isnan(da[0]):
          ldata += [""]
        else:
          ldata += [float(da[0])]
      mdata += [ldata]
    parsed += [mdata]
  file = open(f'../../app/public/data/gistemp/{year}.json', 'w')
  file.write(json.dumps(parsed))

File: InfoVis/app/src/gistemp.js
export class Gistemp {
  constructor() {
    this.data = {};
  }

  async fetch(from, to) {
    if (to === undefined) {
      to = from;
    }
    for (let year = from; year <= to; year++) {
      if (!this.data[year]) {
        const response = await fetch(`/data/gistemp/${year}.json`);
        const raw = await response.json();
        this.data[year] = this._convert(raw);
        console.info(`Fetched temperature data from ${year}`);
      }
    }
  }

  _convert(data) {
    return data.map((d) => {
      const parsed = {};

      for (let i = 0, lat = -89; i <= d.length; i++, lat += 2) {
        const temps = d[i];
        parsed[lat] = {};

        if (!temps) {
          continue;
        }

        for (let i = 0, lon = -179; i <= temps.length; i++, lon += 2) {
          if (temps[i]) {
            parsed[lat][lon] = temps[i];
          }
        }
      }

      return parsed;
    });
  }

  get(year, month) {
    if (month) {
      return this.data[year]?.[month - 1];
    } else {
      return this.data[year];
    }
  }
}
